test/test-applications/integrations/TestApplication.ContinuousProfiler/Exporter/ExtendedPprofBuilder.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using Google.Protobuf;
+using OpenTelemetry.Proto.Common.V1;
+using OpenTelemetry.Proto.Profiles.V1.Alternatives.PprofExtended;
+
+namespace TestApplication.ContinuousProfiler;
+
+internal class ExtendedPprofBuilder
+{
+    private readonly LocationCache _locationCache;
+    private readonly LinkCache _linkCache;
+    private readonly AttributeCache _attributeCache;
+
+    public ExtendedPprofBuilder()
+    {
+        Profile = new Profile();
+        var stringCache = new StringCache(Profile);
+        var functionCache = new FunctionCache(Profile, stringCache);
+        _locationCache = new LocationCache(Profile, functionCache);
+        _linkCache = new LinkCache(Profile);
+        _attributeCache = new AttributeCache(Profile);
+    }
+
+    public Profile Profile { get; }
+
+    public ulong GetLocationId(string function) => _locationCache.Get(function);
+
+    public void AddLink(SampleBuilder sampleBuilder, long spanId, long traceIdHigh, long traceIdLow)
+    {
+        var linkId = _linkCache.GetOrAdd(spanId, traceIdHigh, traceIdLow);
+        sampleBuilder.SetLink(linkId);
+    }
+
+    public void AddAttribute(SampleBuilder sample, string name, string value)
+    {
+        AddAttribute(sample, name, anyValue => anyValue.StringValue = value);
+    }
+
+    public void AddAttribute(SampleBuilder sampleBuilder, string name, long value)
+    {
+        AddAttribute(sampleBuilder, name, anyValue => anyValue.IntValue = value);
+    }
+
+    private void AddAttribute(SampleBuilder sampleBuilder, string name, Action<AnyValue> setValue)
+    {
+        var attributeId = _attributeCache.GetOrAdd(name, setValue);
+
+        sampleBuilder.AddAttribute(attributeId);
+    }
+
+    private class StringCache
+    {
+        private readonly Profile _profile;
+        private readonly Dictionary<string, long> _table = new();
+        private long _index;
+
+        public StringCache(Profile profile)
+        {
+            _profile = profile;
+            GetOrAdd(string.Empty); // 0 is reserved for the empty string
+        }
+
+        public long GetOrAdd(string str)
+        {
+            if (_table.TryGetValue(str, out var value))
+            {
+                return value;
+            }
+
+            _profile.StringTable.Add(str);
```

Comment:
If multiple tests run in parallel for this feature, this could cause concurrency issues. Better to rely on concurrentDictionary GetorAdd.

Reply:
It is purely one-threaded code. There is no possibility to execute exporting process twice or more in the same time (at least for now). It is part of the testing application for now.